geoopt/manifolds/poincare/math.py

```diff
@@ -343,13 +343,12 @@ def mobius_coadd(x, y, *, c=1.0, dim=-1):
 
 
 def _mobius_coadd(x, y, c, dim: int = -1):
-    y = y + 1e-15
     x2 = x.pow(2).sum(dim=dim, keepdim=True)
     y2 = y.pow(2).sum(dim=dim, keepdim=True)
     num = (1 - c * y2) * x + (1 - c * x2) * y
     denom = 1 - c ** 2 * x2 * y2
     # avoid division by zero in this way
-    return num / (denom + 1e-15)
+    return num / denom.clamp_min(1e-15)
```

Comment:
Good. Now one at least doesn't have to doubt whether division by zero might appear because of negative `denom`. And hard threshold should probably produce less artifacts than soft threshold... in intermediate computations.

Still Mobius addition could in theory be generalized to velocities outside the admissible ball (TODO: reference), so I would've probably allowed negative `denom` with hard threshold reducing its absolute value but not changing the sign

Reply:
Not sure if we intend to provide the generalized version. Moreover, having admissible velosities I was able to achieve small negative denom due to numerical precision